docs/source/util.py

```diff
@@ -0,0 +1,174 @@
+
+.. autosummary::
+    :members:
+    :inherited-members:
+    :toctree: api/
+
+    {name}
+
+"""
+    fname = f"temp/{name}_temp.rst"
+    with open(fname, "w") as fp:
+        fp.write(rst_content)
+
+    subprocess.run(["sphinx-autogen", fname, "-o", "output", "-t", "_templates"])
+
+    section = ""
+
+    with open(f"output/{name}.rst") as fp:
+        for line in fp:
+            line = line.strip()
+            if line.startswith(".. autosummary::"):
+                continue
+            if line.startswith(".. rubric:: "):
+                line = line[12:]
+                for title in mapping.keys():
+                    if line.startswith(title):
+                        section = mapping[title]
+                        break
+                continue
+            if not section:
+                continue
+            if line and not line.isspace():
+                getattr(res, section).append(line.rstrip())
+            elif getattr(res, section):  # End of a section
+                section = ""
+
+    return res
+
+
+def generate_section(section: str, content: str):
+    if content:
+        return f"""
+.. rubric:: {section}
+
+.. autosummary::
+    :toctree: api/
+
+    {content}
+
+"""
+    else:
+        return ""
+
+
+def generate(title: str, module: str, classes: Iterable[str]) -> str:
+    import itertools
+
+    if classes:
+        results = [autogen_and_parse_for_info(module, c) for c in classes]
+        names = "\n    =".join(classes)
+        methods = "\n    ".join(itertools.chain.from_iterable(c.methods for c in results))
+        attributes = "\n    ".join(
+            itertools.chain.from_iterable(c.attributes for c in results)
+        )
+        return f"""
+{'='*(len(title)+5)}
+{title}
+{'='*(len(title)+5)}
+
+.. currentmodule:: {module}
+
+.. rubric:: Classes
+
+.. autosummary::
+    :recursive:
+    :toctree: api/
+
+    {names}
+
+{generate_section("Methods", methods)}
+{generate_section("Attributes", attributes)}
+
+"""
+    else:
+        mod = autogen_and_parse_for_info(module)
+        attributes = "\n    ".join(mod.attributes)
+        functions = "\n    ".join(mod.functions)
+        exceptions = "\n    ".join(mod.exceptions)
+        classes = "\n    ".join(mod.classes)
+        return f"""
+{'='*(len(title)+5)}
+{title}
+{'='*(len(title)+5)}
+
+.. automodule:: {module}
+  :noindex:
+
+.. currentmodule:: {module}
+
+{generate_section("Attributes", attributes)}
+{generate_section("Functions", functions)}
+{generate_section("Classes", classes)}
+{generate_section("Exceptions", exceptions)}
+
+"""
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        prog="generate_doc",
+        description="Generate documentation rst for a module or classes inside a module",
+    )
+
+    parser.add_argument(
+        "module", help="The module or the parent module of the classes to be documented"
+    )
+    parser.add_argument("-c", "--classes", nargs="*", help="Classes to be documented")
```

Comment:
I wonder if we should document how the current docs can be generated using this script. Also, let's just name this `doc_generator.py` or something that is clearer than `util` :p

Reply:
This script should be used for reference only. There are parts of the current files that just require manual change. The usage I imagine would be something like: 

- There're some interface changes made to `DataFrame`
- Run `util.py` to get the latest list of methods/attributes.
- Paste the list in our rst files manually.

Reply:
which parts of the current files require manual change? We can discuss in this PR and also should document it

Reply:
From the top of my head:
- In `types.rst`, autosummary will categorize `Geography` and `Variant` to `Attributes` instead of `Classes` like the rest of type definitions.
- Still in `types.rst`, the current script does not support custom rubric names like `Classes/Type Hints`
- If we want to leave any comment/literal markdown in the `rst`, for example my TODO comment in `session.rst`, the script does not preserve these.

Reply:
I see, let's mention them somewhere?

Reply:
Yeah I think it would be useful to document the known caveat. People are lazy (at least I am :p) and will rely on the script more and more without realizing something might not work if not documented.